### DISCUSSION TITLE
What impact has a containerd restart on my running service tasks?

### DISCUSSION THREAD

**Discussion by d630** [moby/moby#46449] [Q&A] [answered]
Hi. I have an active single-node swarm with 10 services on my Debian system. The docker engine runs containerd standalone and is started via systemd service (`ExecStart=/usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock`). What exactly will not be working if containerd is down? Will my running containers also be restarted if I restart containerd?

**Comment by thaJeztah** [comment, marked as answer]
Existing containers will continue to run, which is handled by the containerd shim (which acts as parent process for the container), but _creating_ and _deleting_ containers won't work while containerd is down; you can try this on a test machine;

```bash
# create a container that prints the date every 2 seconds
docker run -d --rm --name mycontainer alpine sh -c 'while true; do date; sleep 2; done'

# restart containerd, and check the container logs
systemctl restart containerd

docker logs mycontainer
Mon Sep 11 11:28:10 UTC 2023
Mon Sep 11 11:28:12 UTC 2023
Mon Sep 11 11:28:14 UTC 2023


# stop containerd and check that it's still running
systemctl stop containerd
docker logs mycontainer
Mon Sep 11 11:28:10 UTC 2023
Mon Sep 11 11:28:12 UTC 2023
Mon Sep 11 11:28:14 UTC 2023
Mon Sep 11 11:28:16 UTC 2023
Mon Sep 11 11:28:18 UTC 2023
Mon Sep 11 11:28:20 UTC 2023

# but creating or starting a new container will fail
docker run -it --rm alpine
docker: Error response from daemon: connection error: desc = "transport: Error while dialing dial unix:///run/containerd/containerd.sock: timeout": unavailable.
```


I should mention that the "remove containers" while containerd is not running situation could use some improvement (in general you'd not want containerd to be stopped for longer periods of time). I'll open a ticket for that to see if we can make that slightly better.

**Reply by thaJeztah** [reply]
> I'll open a ticket for that to see if we can make that slightly better.

I see I forgot to post here; I opened a ticket (still some things to be detailed there, but conversation is happening there on possible improvements);

- https://github.com/moby/moby/issues/46451

**Reply by d630** [reply]
Ah, I see. I suppose I should dive deeper into the shims topic. But for now, that is all I needed to know. Thank you!
